clarifai/runners/models/model_class.py

```diff
@@ -1,41 +1,232 @@
 
-  def predict_wrapper(
-      self, request: service_pb2.PostModelOutputsRequest) -> service_pb2.MultiOutputResponse:
-    """This method is used for input/output proto data conversion"""
-    return self.predict(request)
-
-  def generate_wrapper(self, request: service_pb2.PostModelOutputsRequest
-                      ) -> Iterator[service_pb2.MultiOutputResponse]:
-    """This method is used for input/output proto data conversion and yield outcome"""
-    return self.generate(request)
-
-  def stream_wrapper(self, request: service_pb2.PostModelOutputsRequest
-                    ) -> Iterator[service_pb2.MultiOutputResponse]:
-    """This method is used for input/output proto data conversion and yield outcome"""
-    return self.stream(request)
-
   @abstractmethod
   def load_model(self):
+    """Load the model."""
     raise NotImplementedError("load_model() not implemented")
 
   @abstractmethod
-  def predict(self,
-              request: service_pb2.PostModelOutputsRequest) -> service_pb2.MultiOutputResponse:
-    raise NotImplementedError("run_input() not implemented")
+  def predict(self, *args, **kwargs) -> Output:
+    """Predict method for single or batched inputs."""
+    raise NotImplementedError("predict() not implemented")
 
   @abstractmethod
-  def generate(self, request: service_pb2.PostModelOutputsRequest
-              ) -> Iterator[service_pb2.MultiOutputResponse]:
+  def generate(self, *args, **kwargs) -> Iterator[Output]:
+    """Generate method for streaming outputs."""
     raise NotImplementedError("generate() not implemented")
 
   @abstractmethod
-  def stream(self, request_iterator: Iterator[service_pb2.PostModelOutputsRequest]
-            ) -> Iterator[service_pb2.MultiOutputResponse]:
+  def stream(self, *args, **kwargs) -> Iterator[Output]:
+    """Stream method for streaming inputs and outputs."""
     raise NotImplementedError("stream() not implemented")
+
+  def batch_predict(self, inputs: List[Dict[str, Any]]) -> List[Output]:
+    """Batch predict method for multiple inputs."""
+    with ThreadPoolExecutor() as executor:
+      futures = [executor.submit(self.predict, **input) for input in inputs]
+      return [future.result() for future in futures]
+
+  def batch_generate(self, inputs: List[Dict[str, Any]]) -> Iterator[List[Output]]:
+    """Batch generate method for multiple inputs."""
+    with ThreadPoolExecutor() as executor:
+      futures = [executor.submit(self.generate, **input) for input in inputs]
+      return [future.result() for future in futures]
+
+  def batch_stream(self, inputs: List[Dict[str, Any]]) -> Iterator[List[Output]]:
+    """Batch stream method for multiple inputs."""
+    NotImplementedError("batch_stream() not implemented")
+
+  def predict_wrapper(
+      self, request: service_pb2.PostModelOutputsRequest) -> service_pb2.MultiOutputResponse:
+    outputs = []
+    inputs = self._convert_proto_to_python(request.inputs)
+    if len(inputs) == 1:
+      inputs = inputs[0]
```

Comment:
oh this will be nice as in this wrapper we can handle things that the user should not have to deal with like including in each output the input.id too which some of our APIs / code